cleanlab/outlier.py

```diff
@@ -0,0 +1,541 @@
+        labels: Optional[np.ndarray] = None,
+        params: dict = None,
+        verbose: bool = True,
+    ) -> np.ndarray:
+        """
+        Shared fit functionality between `fit()` and `fit_score()`.
+        For details, refer to :py:func:`fit <cleanlab.outlier.OutOfDistribution.fit>`
+        or :py:func:`fit_score <cleanlab.outlier.OutOfDistribution.fit_score>`.
+        """
+        # Assert valid inputs and params
+        self._assert_valid_inputs(features, pred_probs)
+
+        # Update params
+        self.params = self.DEFAULT_PARAM_DICT
+        if params is not None:
+            self.params = {**self.params, **params}
+
+        if features is not None:
+            # raise Error if they passed in invalid params keys for using features.
+            ood_params = self._get_invalid_params(params, self.OUTLIER_PARAMS)
+            if len(ood_params) > 0:
+                raise ValueError(
+                    f"If fit with features, passed in params dict can only contain {self.OUTLIER_PARAMS}. Remove {ood_params} from params dict."
+                )
+            # get outlier scores
+            if verbose:
+                print("Fitting OOD object based on provided features ...")
+            scores, knn = _get_ood_features_scores(
+                features, **self._get_params(self.OUTLIER_PARAMS), return_estimator=True
+            )
+            self.knn = knn  # save estimator
+
+        if pred_probs is not None:
+            # raise Error if they passed in invalid params keys for using pred_probs.
+            outlier_params = self._get_invalid_params(params, self.OOD_PARAMS)
+            if len(outlier_params) > 0:
+                raise ValueError(
+                    f"If fit with pred_probs, passed in params dict can only contain {self.OOD_PARAMS}. Remove {outlier_params} from params dict."
+                )
+
+            # get ood scores
+            if verbose:
+                print("Fitting OOD object based on provided pred_probs ...")
+            scores, confident_thresholds = _get_ood_predictions_scores(
+                pred_probs,
+                labels=labels,
+                **self._get_params(self.OOD_PARAMS),
+                return_thresholds=True,
+            )
+            if confident_thresholds is None:
+                warnings.warn(
+                    f"Object not fit with confident_thresholds since adjust_pred_probs=False and no "
+                    f"confident_thresholds were calculated.",
+                    UserWarning,
+                )
+            self.confident_thresholds = confident_thresholds
+        return scores
+
+
+def _get_ood_features_scores(
+    features: Optional[np.ndarray] = None,
+    knn: Optional[NearestNeighbors] = None,
+    k: Optional[int] = None,
+    t: int = 1,
+    return_estimator: bool = False,
+) -> Union[np.ndarray, Tuple[np.ndarray, NearestNeighbors]]:
+    """Returns an outlier score for each example based on its feature values.
```

Comment:
which is computed inversely proportional to the average distance between this example and its K nearest neighbors (in feature space).